Laser damage falloff & Laser/Pulse tweaks

Kell-E opened the pull request:
All guns may now optionally have damage falloff using the damage_falloff variable.
Lasers now suffer from damage falloff (Human weaponry only, I've not touched alien weapons. If there's a desire for it I can apply the same changes to them.)

Weapons with damage falloff deal less damage, agony and penetrate less effectively the further they have to travel towards their target. 
Updated: Weapons now use a series of range bands to decide on their final damage falloff rather than relying on math to do the heavy lifting. 

The compact smartguns lethal damage has been increased to thirty fie and the number of shots increased from five to six. The compact smartgun has the most rapid drop in effectiveness of all lasers, when used at close range this should perform as a small buff but when firing at a distance it should be much less effective.

All pulse weapons have had their accuracy tweaked slightly, in addition they've all gained a large armor penetration buff.
Pulse weapons suffer from damage falloff similarly to lasers but to a lesser effect.
I've also retweaked the plasma cutter to be significantly more dangerous, the damage increased from fifteen to thirty and it's armor penetration increased from zero to thirty. It should be able to function as a threatening weapon up close but suffer significantly once any distance is placed between it's user and target as it has a similar damage falloff to the compact smartgun.

The largest concern and the one pointed out to me is that this will promote players trying to get as absolutely close as possible when using their lasers to mitigate the distance penalties potentially ignoring danger in favor of effectiveness. 

🆑 Kell-E
rscadd: All guns may now optionally have damage falloff 
balance: All laser weapons suffer from damage falloff relative to their size
balance: Compact smartgun: Shots (5 ->6), Lethal damage (25 -> 35), Armor penetration (10 -> 0) 
balance: Pulse pistol: Armor penetration (0 -> 25), Accuracy (-1 -> 0)
balance: Pulse carbine: Armor penetration (0 ->30), Accuracy (-1 -> 0)
balance: Pulse rifle: Armor penetration (0 -> 35), Accuracy (-1 -> 1)
balance: Plasma cutter: Damage (15 -> 30), Armor penetration (0 -> 30)
/🆑

SierraKomodo commented on code/modules/projectiles/projectile.dm:
```diff
@@ -29,6 +29,7 @@
 	var/hitchance_mod = 0
 	var/dispersion = 0.0
 	var/distance_falloff = 2  //multiplier, higher value means accuracy drops faster with distance
+	var/damage_falloff = 0 
```
Non-blocking request, but can you add a dmdoc comment explaining what the var does/its intended use? It helps with reviews and future updates.

SierraKomodo commented on code/modules/projectiles/projectile.dm:
```diff
@@ -200,11 +201,17 @@
 	//makes moving targets harder to hit, and stationary easier to hit
 	var/movment_mod = min(5, (world.time - target_mob.l_move_time) - 20)
+
+	var/damage_mod = min(damage_falloff*(distance)*(distance)/100, 0.8)
```
Was multiplying by distance twice intentional or a typo?

Kell-E replied:
Intentional, squaring distance provides progressively punishing penalties the further away the target is. I didn't want the falloff to be particularly notable within the weapons intended range.

Spookerton replied:
To be honest, an algorithmic approach is probably dramatically preferable for balance than any take on exponential math.
```
/// A list of list(distance, multiplier) pairs!
...beam/var/list/falloff

/// Under 10, damage_mod is 1. Equal/Over 16, it's 0.25. Between distances, lower wins.
mybeam/falloff = list(
  list(10, 0.75),
  list(13, 0.5),
  list(16, 0.25)
)


mybeam/DoDamage()
  var/distance = (we know this from somewhere, it's an example)
  var/damage_mod = 1
  for (var/list/entry as anything in falloff)
    if (entry[1] > distance)
      break
    damage_mod = entry[2]
  // damage_mod is now appropriate!
```

Kell-E replied:
Done! Or I'd like to think I've gotten it pretty close to what was described at least.
Values will probably need tweaking because I just went with gut instinct on a lot of things.

SierraKomodo commented on code/modules/projectiles/projectile.dm:
```diff
@@ -200,11 +201,17 @@
 	//makes moving targets harder to hit, and stationary easier to hit
 	var/movment_mod = min(5, (world.time - target_mob.l_move_time) - 20)
+
+	var/damage_mod = min(damage_falloff*(distance)*(distance)/100, 0.8)
+	if (damage_falloff != 0)
+		damage = damage * (1 - damage_mod)
+		armor_penetration = armor_penetration * (1- damage_mod)
+		agony = agony * (1 - damage_mod)
```
Considering the result of `damage_mod` is only used if `damage_falloff` has a value, may as well move that calculation step within the if statement.
```suggestion
	if (damage_falloff != 0)
		var/damage_mod = min(damage_falloff*(distance)*(distance)/100, 0.8)
		damage = damage * (1 - damage_mod)
		armor_penetration = armor_penetration * (1- damage_mod)
		agony = agony * (1 - damage_mod)
```

SierraKomodo commented on code/modules/projectiles/projectile.dm:
```diff
@@ -29,6 +29,12 @@
 	var/dispersion = 0.0
 	var/distance_falloff = 2  //multiplier, higher value means accuracy drops faster with distance
+	var/damage_falloff = FALSE 
+	var/damage_falloff_list = list(
+		list(4, 0.9),
+		list(6, 0.8),
+		list(8, 0.7)
+	) // List(Distance, Multiplier), intended to represent short / medium / long ranges. Uses default of 1 for anything lower than the first value.
```
dmdoc formatting, so it parses properly in all cases.

```suggestion
	/// List(Distance, Multiplier), intended to represent short / medium / long ranges. Uses default of 1 for anything lower than the first value.
	var/damage_falloff_list = list(
		list(4, 0.9),
		list(6, 0.8),
		list(8, 0.7)
	)
```

SierraKomodo reviewed (dismissed):
Looks good on my end now. Just needs squashing.

SierraKomodo reviewed (dismissed):
Merge conflicts

Kell-E commented:
Okay so this branch is now dead too, I'll... redo it all tomorrow

Kell-E commented:
Nevermind the branch now lives, I'll do a quick pass over the values to make sure it all lines up though.

Kell-E commented:
Updated the base values to make the drop-off more tolerable, I have no idea what I was doing with the first batch of values. I'm going with a conservative set of base values and then looking to adjust later if there are any obvious issues with weapons.

SierraKomodo commented on code/modules/projectiles/projectile.dm:
```diff
@@ -200,11 +207,27 @@
 	//makes moving targets harder to hit, and stationary easier to hit
 	var/movment_mod = min(5, (world.time - target_mob.l_move_time) - 5)
+
+	if (damage_falloff == TRUE)
```
` == TRUE` is redundant for boolean values.

```suggestion
	if (damage_falloff)
```

SierraKomodo commented on code/modules/projectiles/projectile.dm:
```diff
@@ -200,11 +207,27 @@
+			if (entry[1] > distance)
+				break
+			damage_mod = entry[2]
+		damage = damage * damage_mod
+		armor_penetration = armor_penetration * damage_mod
+		agony = agony * damage_mod
```
Projectiles that can penetrate and hit multiple mobs may have an unintended side effect of drastically diminishing returns due to direct modification of these vars. While theoretically, that's probably a good thing, there may already be code making subsequent impacts less effective.

Probably not worth digging into here because I imagine that would require a lot more updates to underlying projectile code, but it is worth noting for the future.

Kell-E replied:
I'll have to take a look at it another time then, I wasn't actually sure how projectile penetration factored into things. Fortunately I don't think there's a single projectile that should be effected by this for now.

Kell-E committed:
Laser damage falloff & Laser/Pulse tweaks

All guns may now optionally have damage falloff using the damage_falloff variable.
Lasers now suffer from damage falloff (Human weaponry only, I've not touched alien weapons. If there's a desire for it I can apply the same changes to them at a later date.)

Weapons with damage falloff deal less damage, agony and penetrate less effectively the further they have to travel towards their target. Generally most weapons affected by this have had their base values increased so that they remain similarly effective to three or four tiles depending on their size but will see a decrease to effectiveness when traveling further.

The compact smartguns lethal damage has been increased to thirty fie and the number of shots increased from five to six. The compact smartgun has the most rapid drop in effectiveness of all lasers, when used at close range this should perform as a small buff but when firing at a distance it should be much less effective.

All pulse weapons have had their accuracy tweaked slightly, in addition they've all gained a large armor penetration buff.
Pulse weapons suffer from damage falloff similarly to lasers but to a lesser effect.
I've also retweaked the plasma cutter to be significantly more dangerous, the damage increased from fifteen to thirty and it's armor penetration increased from zero to thirty. It should be able to function as a threatening weapon up close but suffer significantly once any distance is placed between it's user and target as it has a similar damage falloff to the compact smartgun.

The largest concern and the one pointed out to me is that this will promote players trying to get as absolutely close as possible when using their lasers to mitigate the distance penalties potentially ignoring danger in favor of effectiveness. 

Fixed a minor mistake


Changes based upon review


Laser damage falloff take two

Damage falloff values will absolutely need tweaking but I think their initial values settle into a nice place.
Formatting changes


a


Better base falloff values


changes based upon review
